### PR TITLE
Enable Time schedule

### PR DESCRIPTION
**Pull request by ecfm**
Change the existing `TimeScheduler` to publish notification when a scheduled job is triggered. `NotificationSubscriberService` updates job queue accordingly when receiving TIME type notification.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  /**
+   * Checks if all schedule requirements are satisfied,
+   * then publish notification to trigger the corresponding schedule.
+   *
+   * @param programId Program Id
+   * @param systemOverrides Arguments that would be supplied as system runtime arguments for the program.
+   * @param userOverrides Arguments to add to the user runtime arguments for the program.
+   */
+  public void publishNotification(ProgramId programId, Map<String, String> systemOverrides,
+                                 Map<String, String> userOverrides) throws Exception {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: misalignment

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+  private static final Gson GSON = new Gson();
+
+  private final MessagingService messagingService;
+  private final TopicId topicId;
+  private final Store store;
+  private final PropertiesResolver propertiesResolver;
+  private final RunConstraintsChecker requirementsChecker;
+  public ScheduleTaskPublisher(Store store, MessagingService messagingService,
+                            PropertiesResolver propertiesResolver, CConfiguration cConf) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: misalignment

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    userArgs.putAll(userOverrides);
+
+    systemArgs.putAll(propertiesResolver.getSystemProperties(programId.toId()));
+    systemArgs.putAll(systemOverrides);
+
+    Map<String, String> properties = new HashMap<>();
+    properties.put("userArgs", GSON.toJson(userArgs));
+    properties.put("systemArgs", GSON.toJson(systemArgs));
+    properties.put("programId", programId.toString());
+    properties.put("scheduleName", scheduleName);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: The logical time of a the time event should be part of the notification as well.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java [hunk 4]
```diff
@@ -75,20 +82,23 @@ public void process(ApplicationWithPrograms input) throws Exception {
...
-        createSchedule(appId.program(programType, newScheduleSpec.getProgram().getProgramName()), newScheduleSpec);
+        createSchedule(programId, newScheduleSpec);
         continue;
       }
 
-      scheduler.updateSchedule(appId.program(programType, newScheduleSpec.getProgram().getProgramName()),
+      scheduler.updateSchedule(programId,
                                newScheduleSpec.getProgram().getProgramType(),
                                newScheduleSpec.getSchedule());
+      programScheduler.deleteSchedule(input.getApplicationId().schedule(oldScheduleSpec.getSchedule().getName()));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: why is this needed? It means that, when an application is redeployed, we always delete and recreate its program schedules. Even if its schedules have not changed at all?

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: And deleting also means we will cancel all pending jobs in the job queue?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java [hunk 5]
```diff
@@ -103,5 +113,19 @@ public void process(ApplicationWithPrograms input) throws Exception {
 
   private void createSchedule(ProgramId programId, ScheduleSpecification scheduleSpec) throws SchedulerException {
     scheduler.schedule(programId, scheduleSpec.getProgram().getProgramType(), scheduleSpec.getSchedule());
+    addProgramSchedule(programId, scheduleSpec);
+  }
+
+  private void addProgramSchedule(ProgramId programId, ScheduleSpecification scheduleSpec) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: So, now we have two stages: one to create old schedules, one to create new program schedules, and each one may create schedules of the other type. I find that rather confusing in terms of code structure. It appears that a single stage that deals with both types of schedules would have more clarity.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java [hunk 6]
```diff
@@ -103,5 +113,19 @@ public void process(ApplicationWithPrograms input) throws Exception {
...
+    Schedule schedule = scheduleSpec.getSchedule();
+    if (schedule instanceof TimeSchedule) {
+      ProgramSchedule programSchedule =
+        Schedulers.getProgramSchedule((TimeSchedule) schedule, programId);
+      try {
+        programScheduler.addSchedule(programSchedule);
+      } catch (AlreadyExistsException e) {
+        LOG.warn("ProgramSchedule {} already exists for program {}", schedule.getName(), programId);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: It seems that this will now always happen when an app is updated. If its schedules don't change, are we gonna try to create the program schedule and then log this warning? That is misleading because there is really nothing wrong. 

It may be an error if the schedule exists for a different program. Not sure. But then it is probably an error, not a warning.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java [hunk 7]
```diff
@@ -63,10 +65,11 @@ public void process(ApplicationWithPrograms input) throws Exception {
       // delete schedules that existed in the old app spec, but don't anymore
       ScheduleSpecification scheduleSpec = entry.getValue();
       ProgramType programType = ProgramType.valueOfSchedulableType(scheduleSpec.getProgram().getProgramType());
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java:78]: Couple of concerns:
- I thought we wanted to stop doing this (deleting schedules that were removed from the app spec). Because otherwise, if I add a schedule using REST, and then update the app, the schedule will be deleted. 
- What happens if I had an (old-style) time schedule in my app, now I change my app to have the same, but now it is a program schedule with a time trigger? Will that lead to deletion of the schedule?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    String scheduleName = systemOverrides.get(ProgramOptionConstants.SCHEDULE_NAME);
+    ApplicationSpecification appSpec = store.getApplication(programId.getParent());
+    if (appSpec == null) {
+      throw new TaskExecutionException(String.format(UserMessages.getMessage(UserErrors.PROGRAM_NOT_FOUND), programId),
+                                       false);
+    }
+
+    ScheduleSpecification spec = appSpec.getSchedules().get(scheduleName);
+    if (!requirementsChecker.checkSatisfied(programId, spec.getSchedule())) {
+      return;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: why would you check the run constraints here? All you should do it publish. The scheduler's subscriber will receive the notification and create a new job in the job queue. Then it will check the constraints before  the job is launched. 

If you check here, then the job will not ever be launched, even if the constraints are satisfied 5 minutes later. Also, it is better to have the constraint checking in a single place, for separation of concerns

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java (hunk 3 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: I think only the schedule Id and the logical run time should be in the notification. Everything else can be done by the program scheduler's launcher. This is also for decoupling the time triggering from job instantiation logic.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java [hunk 9]
```diff
@@ -41,4 +47,10 @@ public static final String triggerKeyForPartition(DatasetId datasetId) {
                             scheduleId.getVersion(),
                             scheduleId.getSchedule());
   }
+
+  public static ProgramSchedule getProgramSchedule(TimeSchedule timeSchedule, ProgramId programId) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/Schedulers.java]: this is not really a getter but a converter. Maybe call it toProgramSchedule()

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/CoreSchedulerService.java [hunk 10]
```diff
@@ -137,7 +137,7 @@ public Void run(ProgramScheduleStoreDataset store) {
         store.deleteSchedules(appId);
         return null;
       }
-    }, null);
+    }, RuntimeException.class);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/CoreSchedulerService.java:143]: thanks for fixing these

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java [hunk 11]
```diff
@@ -319,15 +334,31 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
   }
 
   private static final class Job {
-    ProgramSchedule schedule;
+    private final ProgramSchedule schedule;
+    private final Map<String, String> systemArgs;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java]: I am not sure why they are needed here. Can't we fetch that from the schedule when the job is launched? Or is the concern that the schedule may be updated/changed/deleted before the job is launched?

**Review by anew — COMMENTED** [review]
Please see comments, thanks -Andreas

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java (hunk 3 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: In `DefaultSchedulerService`, `builder.put(ProgramOptionConstants.RETRY_COUNT, Integer.toString(context.getRefireCount()));` this is also passed in `systemOverrides`. Not sure where this will be used though.

Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: Also in `DefaultSchedulerService`, contents in `trigger.getJobDataMap()` are also passed in as `systemOverrides`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java (hunk 8 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: This is just the existing constraint. Currently there's only number of concurrent-runs constraint. If it's not checked here, I also need to convert existing constraints to new constraints.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateProgramSchedulesStage.java [hunk 12]
```diff
@@ -63,9 +63,7 @@ public void process(final ApplicationWithPrograms input) throws Exception {
       Trigger trigger = scheduleCreationSpec.getTrigger();
       ProgramSchedule programSchedule =
         new ProgramSchedule(scheduleCreationSpec.getName(), scheduleCreationSpec.getDescription(),
-                            programId,
-                            scheduleCreationSpec.getProperties(), trigger,
-                            new ArrayList<Constraint>());
+                            programId, scheduleCreationSpec.getProperties(), trigger, new ArrayList<Constraint>());
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateProgramSchedulesStage.java]: Should this call ScheduleCreationSpec.getConstraints()? Even if constraints are not implemented yet, once they are, this will have to change.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java (hunk 4 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: perhaps we should have a updateSchedule() in Scheduler, that internally deletes and adds. That way, we don't hardcode here, how update is done.

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: let's do that in a separate PR. https://issues.cask.co/browse/CDAP-11494

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java (hunk 5 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java]: Let's address that separately: https://issues.cask.co/browse/CDAP-11495

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java [hunk 13]
```diff
@@ -69,7 +70,9 @@ public void process(ApplicationWithPrograms input) throws Exception {
       scheduler.deleteSchedule(input.getApplicationId().program(programType,
                                                                 scheduleSpec.getProgram().getProgramName()),
                                scheduleSpec.getProgram().getProgramType(), scheduleSpec.getSchedule().getName());
-      programScheduler.deleteSchedule(input.getApplicationId().schedule(scheduleSpec.getSchedule().getName()));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java]: I can't comment on line 55 in this PR as there was no change. But: if a program is removed from the app, we need to delete its schedules, regardless of canUpdateSchedules().

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java [hunk 14]
```diff
@@ -68,7 +57,7 @@ public void execute(JobExecutionContext context) throws JobExecutionException {
       String key = trigger.getKey().getName();
       String[] parts = key.split(":");
       Preconditions.checkArgument(parts.length == 6, String.format("Trigger's key name %s has %d parts instead of 6",
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java:60]: Isn't that key a scheduleId? So you can use ScheduleId.fromNameParts()

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java [hunk 15]
```diff
@@ -51,55 +48,34 @@
...
-    systemArgs.putAll(propertiesResolver.getSystemProperties(programId.toId()));
-    systemArgs.putAll(systemOverrides);
-
     Map<String, String> properties = new HashMap<>();
-    properties.put("userArgs", GSON.toJson(userArgs));
-    properties.put("systemArgs", GSON.toJson(systemArgs));
     properties.put("programId", programId.toString());
-    properties.put("scheduleName", scheduleName);
+    properties.put(ProgramOptionConstants.SCHEDULE_NAME, scheduleName);
+    properties.put(ProgramOptionConstants.RETRY_COUNT, Integer.toString(retryCount));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: what is this used for?

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java [hunk 16]
```diff
@@ -304,12 +308,25 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
...
+          throw new ApplicationNotFoundException(scheduleId.getParent());
+        }
+        ScheduleSpecification scheduleSpec = appSpec.getSchedules().get(scheduleId.getSchedule());
+        if (scheduleSpec == null) {
+          throw new NotFoundException(scheduleId);
+        }
+        schedule = Schedulers.toProgramSchedule((TimeSchedule) scheduleSpec.getSchedule(), programId,
+                                                scheduleSpec.getProperties());
+      }
+      addJob(new Job(schedule, ImmutableList.of(notification)));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java]: This will be changed by @anwar6953 's PR, right?

**Review by anew — CHANGES_REQUESTED** [review]
Still a few comments.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java (hunk 13 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java]: right, I'll add this in `ApplicationRegistrationStage`, that's where schedules with deleted programs are removed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java (hunk 14 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java:60]: the separator here is ":" but ScheduleId.fromIdParts() parses with separator "." I'll add a helper method in AbstractSchedulerService

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java:60]: why are we using different separators?

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java:60]: actually fromIdParts() takes the parts, it does not parse.

**Comment by anew** [comment]
Can you please remove schedule migration from this PR and open a new one for that?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java [hunk 17]
```diff
@@ -118,7 +132,7 @@ public boolean apply(Map.Entry<String, ScheduleSpecification> input) {
                                                newSpecification.getFlows(), newSpecification.getMapReduce(),
                                                newSpecification.getSpark(), newSpecification.getWorkflows(),
                                                newSpecification.getServices(), filteredExistingSchedules,
-                                               newSpecification.getProgramSchedules(),
+                                               filteredExistingProgramSchedules,
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ApplicationRegistrationStage.java]: this was not the point of my comment. The new app spec must be consistent in that it does not define schedules for programs that don't exist. I assume that is checked somewhere else. Which means that this filetring won't filter anything out. 

However, somewhere you need to delete the schedules from the schedule store. In the ScheduleDeletionStage you will not do that if canUpdateSchedules is false. That was original comment: even if it is false, it needs to delete these program schedules. Not from the app spec, but from the scheduler.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java (hunk 13 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/DeleteScheduleStage.java]: that is not the right fix. see my comment on your new commit.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java (hunk 15 above)
Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: can you answer my question?

**Review by anew — COMMENTED** [review]
see my comment

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java (hunk 15 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskPublisher.java]: Not sure... the original TimeSchedule run programs with this as one of the runtime args. I just kept it in case it's useful. Haven't found any usage of ProgramOptionConstants.RETRY_COUNT in any other places though

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java (hunk 14 above)
Comment (ecfm) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DefaultSchedulerService.java:60]: The key is actually not ScheduleId, it's a ProgramId ending with a schedule name. Also the programType here is a String generated from `SchedulableProgramType` instead from  `ProgramType`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java [hunk 18]
```diff
@@ -273,11 +272,12 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
         // TODO: (CDAP-11469) No need to check appSpec once migration from DatasetBasedTimeScheduleStore is done
         ApplicationSpecification appSpec = store.getApplication(scheduleId.getParent());
         if (appSpec == null) {
-          throw new ApplicationNotFoundException(scheduleId.getParent());
+          LOG.debug("Cannot find application '{}' for schedule '{}' in AppMetadataStore",
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java]: Is this an indication of a serious inconsistency? or is this expected to happen? Under which circumstances? If iyt is serious, it should log a warning.

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java [hunk 19]
```diff
@@ -273,11 +272,12 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
...
+                    scheduleId.getParent(), scheduleId);
         }
         ScheduleSpecification scheduleSpec = appSpec.getSchedules().get(scheduleId.getSchedule());
         if (scheduleSpec == null) {
-          throw new NotFoundException(scheduleId);
+          LOG.debug("Cannot find schedule '{}' in application '{}'", scheduleId.getSchedule(), scheduleId.getParent());
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java]: same here

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java [hunk 20]
```diff
@@ -273,11 +272,12 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
         // TODO: (CDAP-11469) No need to check appSpec once migration from DatasetBasedTimeScheduleStore is done
         ApplicationSpecification appSpec = store.getApplication(scheduleId.getParent());
         if (appSpec == null) {
-          throw new ApplicationNotFoundException(scheduleId.getParent());
+          LOG.debug("Cannot find application '{}' for schedule '{}' in AppMetadataStore",
+                    scheduleId.getParent(), scheduleId);
         }
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java]: return? Otherwise you will throw an NPE in the next line.

File: cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java [hunk 21]
```diff
@@ -273,11 +272,12 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
...
         ScheduleSpecification scheduleSpec = appSpec.getSchedules().get(scheduleId.getSchedule());
         if (scheduleSpec == null) {
-          throw new NotFoundException(scheduleId);
+          LOG.debug("Cannot find schedule '{}' in application '{}'", scheduleId.getSchedule(), scheduleId.getParent());
         }
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java]: and return here

**Review by anew — COMMENTED** [review]
Please see my comments. Otherwise it appears that the logic is correct. 

With the current code structure, it is extremely hard to argue whether code is correct or not... I would like it a lot more if all APIs would use the new ProgramSchedule and Scheduler APIs. And all interactions are only in terms of that. That would mean:
- there is only one Schedules stage in the deployment pipeline
- we have utility methods to convert old schedules to new style and back; app spec has a convenience method to retrieve all schedules as new style
- the CoreSchedulerService is the only service we interact with. It has an instance of the old scheduler, and for time and stream schedules, it delegates to that. But all add/update/delete goes through the new APIs.

That would greatly simplify the code, and I would feel a lot more comfortable that it is correct.

Can we do that in a separate PR? 

Please address the few comments I had, rebase and merge.

**Comment by anew** [comment]
LGTM, please merge this.
